openedx/core/djangoapps/content/search/documents.py

```diff
@@ -0,0 +1,215 @@
+"""
+Utilities related to indexing content for search
+"""
+from __future__ import annotations
+import hashlib
+import logging
+
+from django.utils.text import slugify
+from opaque_keys.edx.keys import UsageKey, LearningContextKey
+
+from openedx.core.djangoapps.content_libraries import api as lib_api
+from openedx.core.djangoapps.content_tagging import api as tagging_api
+from openedx.core.djangoapps.xblock import api as xblock_api
+
+log = logging.getLogger(__name__)
+STUDIO_INDEX_NAME = "studio_content"
+
+
+class Fields:
+    """
+    Fields that exist on the documents in our search index
+    """
+    # Meilisearch primary key. String.
+    id = "id"
+    usage_key = "usage_key"
+    type = "type"  # DocType.course_block or DocType.library_block (see below)
+    block_id = "block_id"  # The block_id part of the usage key. Sometimes human-readable, sometimes a random hex ID
+    display_name = "display_name"
+    block_type = "block_type"
+    context_key = "context_key"
+    org = "org"
+    # breadcrumbs: an array of {"display_name": "..."} entries. First one is the name of the course/library itself.
+    # After that is the name of any parent Section/Subsection/Unit/etc.
+    # It's a list of dictionaries because for now we just include the name of each but in future we may add their IDs.
+    breadcrumbs = "breadcrumbs"
+    # tags (dictionary)
+    # See https://blog.meilisearch.com/nested-hierarchical-facets-guide/
+    # and https://www.algolia.com/doc/api-reference/widgets/hierarchical-menu/js/
+    # For details on the format of the hierarchical tag data.
+    tags = "tags"
+    tags_taxonomy = "taxonomy"
+    tags_level0 = "level0"
+    tags_level1 = "level1"
+    tags_level2 = "level2"
+    tags_level3 = "level3"
+    # The "content" field is a dictionary of arbitrary data, depending on the block_type.
+    # It comes from each XBlock's index_dictionary() method (if present) plus some processing.
+    # Text (html) blocks have an "html_content" key in here, capa has "capa_content" and "problem_types", and so on.
+    content = "content"
+
+
+class DocType:
+    """
+    Values for the 'type' field on each doc in the search index
+    """
+    course_block = "course_block"
+    library_block = "library_block"
+
+
+def _meili_id_from_opaque_key(usage_key: UsageKey) -> str:
+    """
+    Meilisearch requires each document to have a primary key that's either an
+    integer or a string composed of alphanumeric characters (a-z A-Z 0-9),
+    hyphens (-) and underscores (_). Since our opaque keys don't meet this
+    requirement, we transform them to a similar slug ID string that does.
+    """
```

Comment:
Yes. I don't really like having to generate this arbitrary ID here. We don't really use it for anything, so it's not a big deal, but it would be nicer to use an ID that's the same as other parts of the system, which would be the case with PublishableEntity's ID/UUID. So I'd happily change this to use that once it's available for courseware.

Reply:
Added a comment to that effect.

https://github.com/openedx/edx-platform/blob/ec0c4a875d7bc4766ff02c50a5e36d8ec6cd3237/openedx/core/djangoapps/content/search/documents.py#L72-L73